metals/src/main/scala/scala/meta/internal/bsp/ConnectionBspStatus.scala

```diff
@@ -13,25 +16,83 @@ class ConnectionBspStatus(
     bspStatus: BspStatus,
     folderPath: AbsolutePath,
     icons: Icons,
-) {
-  private val isServerResponsive = new AtomicBoolean(false)
-  val status: MetalsStatusParams => Unit = bspStatus.status(folderPath, _)
-
-  def connected(serverName: String): Unit =
-    if (isServerResponsive.compareAndSet(false, true))
-      status(ConnectionBspStatus.connectedParams(serverName, icons))
-  def noResponse(serverName: String): Unit =
-    if (isServerResponsive.compareAndSet(true, false)) {
-      scribe.debug("server liveness monitor detected no response")
-      status(ConnectionBspStatus.noResponseParams(serverName, icons))
-    }
+)(implicit rc: ReportContext) {
+  private val status = new AtomicReference[BspStatusState](
+    BspStatusState(Disconnected, None, None)
+  )
+  private val currentSessionErrors = new AtomicReference[Set[String]](Set())
+
+  def connected(serverName: String): Unit = changeState(Connected(serverName))
+
+  def noResponse(): Unit = {
+    scribe.debug("server liveness monitor detected no response")
+    changeState(NoResponse)
+  }
 
   def disconnected(): Unit = {
-    isServerResponsive.set(false)
-    status(ConnectionBspStatus.disconnectedParams)
+    currentSessionErrors.set(Set.empty)
+    changeState(Disconnected)
+  }
+
+  def showError(message: String, pathToReport: Path): Unit = {
+    val updatedSet =
+      currentSessionErrors.updateAndGet(_ + pathToReport.toUri().toString())
+    changeState(ErrorMessage(message), updatedSet)
+  }
+
+  def onReportsUpdate(): Unit = {
+    val currentState = status.get()
+    currentState.currentState match {
+      case error @ ErrorMessage(_) => showState(error, currentState)
+      case _ =>
+    }
+  }
+
+  def isBuildServerResponsive: Option[Boolean] =
+    status.get().currentState match {
+      case NoResponse => Some(false)
+      case Disconnected => None
+      case _ => Some(true)
+    }
+
+  private def changeState(
+      newState: BspServerState,
+      errorReports: Set[String] = currentSessionErrors.get(),
+  ) = {
+    val old = status.getAndUpdate(_.changeState(newState)._2)
+    val (newServerState, newBspState) = old.changeState(newState)
```

Comment:
I think this can get confusing, maybe state could have a field that says if something should be shown to a user?